### DISCUSSION TITLE
Running Puma within a container

### DISCUSSION THREAD

**Discussion by ayushn21** [socketry/async-container#36] [Q&A]
This is a follow-up to https://github.com/socketry/async-container/discussions/34.

What's the best way (or is it not possible or not recommended?) to run a Puma server within an async container?

Some context as to why I'm doing this: 
- Largely academic, to see if I can.
- In the future I plan to use Falcon which removes this issue, but all my current projects use Puma and I'd like to see if I can use `async-container` with those.
- I'd like to blog about a way for existing Ruby/Rails users to use `async-container`.

The two problems I'm encountering are:
- The container isn't notified of readiness preventing restarts using `SIGHUP`. As discussed in https://github.com/socketry/async-container/discussions/34, we need to manually notify the container that the process is ready. I don't have control over Puma's executable, so not sure how to do this? 
- If I set `ready: true` when calling `instance.exec` to work around the above issue, sending `SIGHUP` crashes the process because the new process is trying to bind to the same UNIX socket as the existing process throwing an error. Again not sure what the best way to work around this is?

## Code

Here's a simple prototype I threw together for this.  Amend the UNIX socket path in `start` and then run `bundle && bundle exec start`. 

[async-container-puma.zip](https://github.com/user-attachments/files/18210313/async-container-puma.zip)

**Comment by ioquatix** [comment]
I have added an example here showing how to do it: https://github.com/socketry/async-container/tree/main/examples/puma

Can you let me know if this works for you?

**Reply by ayushn21** [reply]
Thanks for this @ioquatix! I didn't know Puma had an `on_booted` hook I could use.

A couple of follow up questions on this:
- The server starts fine now, but restarts using `SIGHUP` still don't work. 
If it's  bound to a port, I get the error:
`Address already in use - bind(2) for "0.0.0.0" port 9292 (Errno::EADDRINUSE)`.
If I bind to a UNIX socket, I get the error:
`There is already a server bound to: /Users/Ayush/Developer/Prototypes/puma.sock (RuntimeError)`

This isn't strictly in the scope of `async-container`, but curious to know if there's a solution to this? We can't have two Puma parent processes listening on the same port/socket can we? What would be the best way to handle this in your opinion? 

- This is more of a hypothetical question. How would you recommend using `async-container` when running arbitrary executables. For eg, if we're running a binary that's not even in Ruby and that doesn't have any lifecycle hooks, how would we go about that because we wouldn't have a way of notifying `async-container` from within the process. Or is this not something we should use `async-container` for?
